### PR TITLE
[training migration] Add CheckpointConfig dataclass

### PR DESCRIPTION
**Pull request by maanug-nv**
# What does this PR do ?

1. Create dataclass in MLM
2. Get MLM arguments to use dataclass

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/arguments.py [hunk 1]
```diff
@@ -2432,131 +2432,34 @@ def _add_learning_rate_args(parser):
...
-    group.add_argument('--save', type=str, default=None,
-                       help='Output directory to save checkpoints to.')
-    group.add_argument('--save-interval', '--persistent-save-interval', type=int, default=None,
-                       help='Number of iterations between persistent checkpoint saves.')
-    group.add_argument('--save-retain-interval', type=int, default=None,
-                       help='Number of iterations between retained checkpoints (other'
-                       'checkpoints _except the last checkpoint_ are automatically deleted).')
+    from megatron.training.config import CheckpointConfig
+
+    ckpt_factory = ArgumentGroupFactory(CheckpointConfig, exclude=["most_recent_k", "save_tokenizer_assets", "save_optim", "save_rng", "load_optim", "load_rng"])
```
Comment (maanug-nv) [inline megatron/training/arguments.py:2516]: `most_recent_k` and `save_tokenizer_assets` are features in Megatron Bridge that are not available in MegatronLM currently.

Comment (maanug-nv) [reply megatron/training/arguments.py:2516]: save/load optim/rng flags are different in bridge and MLM. 
in bridge, save optim is referenced by `CheckpointConfig.save_optim` and is default True. 
in MLM, save optim is referenced by `args.no_save_optim` and is default False.

Not sure how to address this discrepancy right now.

Comment (Wohox) [reply megatron/training/arguments.py:2516]: 1. Can you elaborate more on the discrepancy, the default value of save_optim is aligned in CheckpointConfig and args, both are set to True. The issue is when user adds `--no-save-optim` flag, there is no mapping to CheckpointConfig.
2. I think this PR should replace `self.config.no_save_optim` to `self.config.save_optim`, which won't influence the user args.

This is same for `--no-save-rng --no-load-optim --no-load-rng --no-initialization`

File: megatron/training/arguments.py [hunk 2]
```diff
@@ -2591,131 +2591,34 @@ def _add_learning_rate_args(parser):
...
-    group.add_argument('--use-mp-args-from-checkpoint-args', action='store_true',
-                       help='Copy model parallelism command-line arguments from checkpoint')
-    group.add_argument('--no-use-tokenizer-model-from-checkpoint-args', action='store_false',
-                       dest='use_tokenizer_model_from_checkpoint_args',
-                       help='If set, do not use tokenizer model path from checkpoint')
-    group.add_argument('--exit-on-missing-checkpoint', action='store_true',
-                       help="If '--load' is set, but checkpoint is not found "
-                       "(e.g., path typo), then exit instead of random "
-                       "initialization.")
     group.add_argument('--use-dist-ckpt', action='store_true',
```
Comment (Wohox) [inline megatron/training/arguments.py:2532]: Why does this flag get kept?

Comment (maanug-nv) [reply megatron/training/arguments.py:2532]: it's deprecated. I don't really see a reason to add already deprecated flags to the configs. 
If you have a strong opinion on this, I can add it to the config, it won't break Megatron Bridge.

File: megatron/training/arguments.py (hunk 1 above)
Comment (maanug-nv) [reply megatron/training/arguments.py:2516]: Confused by 2. `CheckpointConfig` currently contains `save_optim`, not `no_save_optim`. 
The basic issue is that `--no-save-optim` and others don't follow the more common pattern of setting `dest="save_optim"`. `ArgumentGroupFactory` is built around that more common pattern. In other words, for boolean settings that are default False, the factory always generates `--no-some-feature` with `dest=some_feature`. save/load optim/rng use `dest=no_save_optim`. 

The best solution is to replace all `args.no_save_optim` with `not args.save_optim`. I also think it makes more sense to do this when we start using the config classes in functions instead of args.

Comment (Wohox) [reply megatron/training/arguments.py:2516]: > The best solution is to replace all args.no_save_optim with not args.save_optim. I also think it makes more sense to do this when we start using the config classes in functions instead of args.

I meant the user should still use `--no-save-optim` flag and the `args.no_save_optim` should be depreacted and value get mapped to `CheckpointConfig.save_optim` (like in #2896: `add_bias_linear: bool = field(
        default=True, metadata={"argparse_meta": {"arg_names": ["--disable-bias-linear"]}}`), then let's replace all the reference of `args/config.no_save_optim` to `not config/args.save_optim`.
